kv/concurrency: batch intent resolution of pushed intents from same txn

nvb opened the pull request:
Fixes #103126.

This commit extends the infrastructure introduced in #49218 for transaction timestamp pushes. It avoids redundant txn pushes of PENDING transactions and batches the resolution of PENDING intents. This breaks the O(num_intents) work performed by high-priority scans (e.g. backups) over intent-heavy keyspaces into something closer to O(num_ranges) work.

The commit accomplishes its goals by adding a second per-Range LRU cache of transactions that are PENDING and are known to have been pushed to higher timestamps. We use this cache for two purposes:

1. when we are a non-locking read and we see a lock at a conflicting timestamp who is held by a pushed txn above our read timestamp, we neither wait out the kv.lock_table.coordinator_liveness_push_delay (50 ms) nor push the transactions record (RPC to leaseholder of pushee's txn record range).
2. we use the existence of a transaction in the cache as an indication that it may have written multiple intents, so we begin deferring intent resolution to enable batching.

Together, these two changes make us much more effective at pushing transactions with a large number of intents. The following example (from #103126) demonstrates this:
```sql
-- SETUP: run in a 3-node GCP roachprod cluster

--- session 1 - write 100k intents
CREATE TABLE keys (k BIGINT NOT NULL PRIMARY KEY);
BEGIN; INSERT INTO keys SELECT generate_series(1, 100000);

--- session 2 - push intents with high-priority txn without uncertainty interval
BEGIN PRIORITY HIGH AS OF SYSTEM TIME '-1ms';
SELECT count(*) FROM keys;

--- BEFORE this PR and before #103265 (i.e. v23.1.2): takes ~7.1ms per intent
Time: 714.441s total

--- BEFORE this PR: takes ~1.5ms per intent
Time: 151.880s total

--- AFTER this PR: takes ~24μs per intent
Time: 2.405s
```

The change does have an unfortunate limitation. Deferred intent resolution is only currently enabled for non-locking readers without uncertainty intervals. Readers with uncertainty intervals must contend with the possibility of pushing a conflicting intent up into their uncertainty interval and causing more work for themselves, which is avoided with care by the lockTableWaiter but difficult to coordinate through the txnStatusCache. This limitation is acceptable because the most important case here is optimizing the Export requests issued by backup.

This limitation also hints at the long-term plan for this interaction, which is that non-locking readers can ignore known pending intents without the need to even resolve those intents (see #94730). This will require a request-scoped cache of pending, pushed transactions, which does not have the same problems with uncertainty intervals.

Release note (performance improvement): Backups no longer perform work proportional to the number of pending intents that they encounter, so they are over 100x faster when encountering long-running, bulk writing transactions.

arulajmani committed:
concurrency: thread cluster settings into the lock table

We'll make use of these when constructing lock modes.

Epic: none

Release note: None

nvb committed:
kv: replace TxnMeta with Transaction in lockTableGuardImpl

Small refactor.

nvb committed:
kv/concurrency: batch intent resolution of pushed intents from same txn

Fixes #103126.

This commit extends the infrastructure introduced in #49218 for transaction
timestamp pushes. It avoids redundant txn pushes of PENDING transactions and
batches the resolution of PENDING intents. This breaks the O(num_intents) work
performed by high-priority scans (e.g. backups) over intent-heavy keyspaces into
something closer to O(num_ranges) work.

The commit accomplishes its goals by adding a second per-Range LRU cache of
transactions that are PENDING and are known to have been pushed to higher
timestamps. We use this cache for two purposes:

1. when we are a non-locking read and we see a lock at a conflicting timestamp
   who is held by a pushed txn above our read timestamp, we neither wait out the
   kv.lock_table.coordinator_liveness_push_delay (50 ms) nor push the
   transactions record (RPC to leaseholder of pushee's txn record range).
2. we use the existence of a transaction in the cache as an indication that
   it may have written multiple intents, so we begin deferring intent resolution
   to enable batching.

Together, these two changes make us much more effective at pushing transactions
with a large number of intents. The following example (from #103126) demonstrates
this:
```sql
-- SETUP: run in a 3-node GCP roachprod cluster

--- session 1 - write 100k intents
CREATE TABLE keys (k BIGINT NOT NULL PRIMARY KEY);
BEGIN; INSERT INTO keys SELECT generate_series(1, 100000);

--- session 2 - push intents with high-priority txn without uncertainty interval
BEGIN PRIORITY HIGH AS OF SYSTEM TIME '-1ms';
SELECT count(*) FROM keys;

--- BEFORE this PR and before #103265 (i.e. v23.1.2): takes ~7.1ms per intent
Time: 714.441s total

--- BEFORE this PR: takes ~1.5ms per intent
Time: 151.880s total

--- AFTER this PR: takes ~24μs per intent
Time: 2.405s
```

The change does have an unfortunate limitation. Deferred intent resolution
is only currently enabled for non-locking readers without uncertainty
intervals. Readers with uncertainty intervals must contend with the
possibility of pushing a conflicting intent up into their uncertainty
interval and causing more work for themselves, which is avoided with care
by the lockTableWaiter but difficult to coordinate through the
txnStatusCache. This limitation is acceptable because the most important
case here is optimizing the Export requests issued by backup.

This limitation also hints at the long-term plan for this interaction,
which is that non-locking readers can ignore known pending intents without
the need to even resolve those intents (see #94730). This will require a
request-scoped cache of pending, pushed transactions, which does not have
the same problems with uncertainty intervals.

Release note (performance improvement): Backups no longer perform work
proportional to the number of pending intents that they encounter, so they are
over 100x faster when encountering long-running, bulk writing transactions.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/104784)
<!-- Reviewable:end -->

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo comments + our discussion from yesterday to not call into `tryUpdateLockLocked` from `resumeScan` for lock updates that correspond to pushed transactions. 

Quite excited to see this perf improvement land! Thanks for doing it :100: 

> Reviewed 2 of 2 files at r1, 14 of 14 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104784#-:-NY43PuGAj48ANvB2ZqU:buqgnko)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/kv/kvserver/concurrency/concurrency_control.go` line 729 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104784#-NY3lXT71a-Z3Po1AvJ3:-NY3lXT71a-Z3Po1AvJ4:b-lhun3l) ([raw file](https://github.com/cockroachdb/cockroach/blob/a89775b23851a8358b668006fd176dce1f3fd435/pkg/kv/kvserver/concurrency/concurrency_control.go#L729)):*
> ```Go
> 	// This is used by the lock table in a best-effort manner to avoid waiting on
> 	// locks of finalized or pushed transactions and telling the caller via
> 	// lockTableGuard.ResolveBeforeEvaluation to resolve a batch of intents.
> ```

This comment seems stale -- we've since renamed the method to be `ResolveBeforeScanning`. Should we update it while we're here?
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1710 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104784#-NY3hqHZBsLbPi2LPkea:-NY3hqHZBsLbPi2LPkeb:bs6pcuy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a89775b23851a8358b668006fd176dce1f3fd435/pkg/kv/kvserver/concurrency/lock_table.go#L1710)):*
> ```Go
> 		// opportunities.
> 		//
> 		// This fast-path is only enabled for readers without uncertainty intervals,
> ```

For readers that do have uncertainty intervals, would it be correct to use `g.ts.Forward(g.txn.GlobalUncertaintyLimit)` instead? 

I realize it would not be as precise as taking local limits into consideration, but is there benefit to doing so instead of completely skipping this optimization? 
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1722 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104784#-NY3iqy63FCmqk6HWoe3:-NY3iqy63FCmqk6HWoe4:b-pts7xo) ([raw file](https://github.com/cockroachdb/cockroach/blob/a89775b23851a8358b668006fd176dce1f3fd435/pkg/kv/kvserver/concurrency/lock_table.go#L1722)):*
> ```Go
> 				up := roachpb.MakeLockUpdate(pushedTxn, roachpb.Span{Key: l.key})
> 				if l.holder.holder[lock.Replicated].txn == nil {
> 					// Only held unreplicated. Update lock directly in case other
> ```

Noting that once non-locking readers stop blocking on `Exclusive` locks, we should be able to guarantee that there is an intent on the key if we ever hit this fast path. At that point, we should be able to assert that the lock is not only held as unreplicated. 

Maybe worth capturing in a TODO, while we have all this paged in?
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1724 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104784#-NY3nsFa439QdcLMWwlJ:-NY3nsFa439QdcLMWwlK:blj2bc2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a89775b23851a8358b668006fd176dce1f3fd435/pkg/kv/kvserver/concurrency/lock_table.go#L1724)):*
> ```Go
> 					// Only held unreplicated. Update lock directly in case other
> 					// waiting readers can benefit from the pushed timestamp.
> 					_, _ = l.tryUpdateLockLocked(up)
> ```

Following on from our conversation yesterday, where we decided we wouldn't call `tryUpdateLockLocked` from `resumeScan` for pushed transactions (we'll only do so for finalized transactions). 

I don't think we want the same behaviour for unreplicated locks -- that's because they don't trigger resolve intent requests, so they won't be calling into `UpdateLocks` as a result. Mostly a note to myself that this should compose well with https://github.com/cockroachdb/cockroach/pull/104782. 
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 2858 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104784#-NY3jZWb4GIJqw8BYYIi:-NY3jZWb4GIJqw8BYYIj:bj7y8u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/a89775b23851a8358b668006fd176dce1f3fd435/pkg/kv/kvserver/concurrency/lock_table.go#L2858)):*
> ```Go
> 		}
> 
> 		if str == lock.None && !g.hasUncertaintyInterval() {
> ```

Consider adding a comment that explains this `g.hasUncertaintyInterval` check (or points to the comment in `tryActiveWait`).
___
*[`pkg/kv/kvserver/concurrency/testdata/concurrency_manager/resolve_pushed_intents` line 95 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104784#-NY409jU805vPZQNdTAN:-NY409jU805vPZQNdTAO:b-8ghwn2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a89775b23851a8358b668006fd176dce1f3fd435/pkg/kv/kvserver/concurrency/testdata/concurrency_manager/resolve_pushed_intents#L95)):*
> ```
> [4] sequence req1: lock wait-queue event: done waiting
> [4] sequence req1: conflicted with ‹00000002-0000-0000-0000-000000000000› on ‹"a"› for 0.000s
> [4] sequence req1: resolving a batch of 9 intent(s)
> ```

This would have made a nice before and after :100: 
___
*[`pkg/kv/kvserver/concurrency/testdata/concurrency_manager/resolve_pushed_intents_without_adding_to_lock_table` line 224 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104784#-NY432YF74y94C9UqO9Y:-NY432YF74y94C9UqO9Z:b-a3rmm5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a89775b23851a8358b668006fd176dce1f3fd435/pkg/kv/kvserver/concurrency/testdata/concurrency_manager/resolve_pushed_intents_without_adding_to_lock_table#L224)):*
> ```
> 
> sequence req=req2
> ----
> ```

Guess this is the before and after I was asking for :joy: 
___
*[`pkg/kv/kvserver/concurrency/testdata/lock_table/resolve_pushed_txn_locks` line 125 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104784#-NY4-Ah886R45ODeAtnE:-NY4-Ah886R45ODeAtnF:b-khdk2z) ([raw file](https://github.com/cockroachdb/cockroach/blob/a89775b23851a8358b668006fd176dce1f3fd435/pkg/kv/kvserver/concurrency/testdata/lock_table/resolve_pushed_txn_locks#L125)):*
> ```
> 
> print
> ----
> ```

Once we implement the changes to not call into `tryUpdateLock` from `resumeScan` for lock updates that correspond to unfinalized transactions, we should see the lock on c back here as well, right?


<!-- Sent from Reviewable.io -->

nvb committed:
kv: gate batched resolution of pushed locks behind a cluster setting

This commit adds a new `kv.lock_table.batch_pushed_lock_resolution.enabled` cluster
setting which controls whether the lock table should allow non-locking readers
to defer and batch the resolution of conflicting locks whose holder is known to
be pending and have been pushed above the reader's timestamp.

This is a safeguard against bugs or behavior changes as we quickly backport a
fix for #103126.

Epic: None
Release note: None

nvb reviewed (commented):
TFTR!

I've also added a cluster setting (in a new commit) so that we can disable this behavior online if we find that it's a problem. Doing so required me to plumb in cluster settings into the lock table, so I cherry-picked the third commit from #104620 into this PR.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104784#-:-NYJnKJaDZSEZLSb9s6O:bks97zi)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani)

___
*[`pkg/kv/kvserver/concurrency/concurrency_control.go` line 729 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104784#-NY3lXT71a-Z3Po1AvJ3:-NYJnJhI9U1Ats62gNER:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a89775b23851a8358b668006fd176dce1f3fd435/pkg/kv/kvserver/concurrency/concurrency_control.go#L729)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

This comment seems stale -- we've since renamed the method to be `ResolveBeforeScanning`. Should we update it while we're here?
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1710 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104784#-NY3hqHZBsLbPi2LPkea:-NYJnjKl7EtQNoS7yGyg:bjmg4h4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a89775b23851a8358b668006fd176dce1f3fd435/pkg/kv/kvserver/concurrency/lock_table.go#L1710)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

For readers that do have uncertainty intervals, would it be correct to use `g.ts.Forward(g.txn.GlobalUncertaintyLimit)` instead? 

I realize it would not be as precise as taking local limits into consideration, but is there benefit to doing so instead of completely skipping this optimization? 
</blockquote></details>

This used to be enough, though we had to fight to keep the pushed timestamp below `hlc.Now()` to avoid pushing the time into the future, needing to mark it as synthetic, and causing the pushee to commit-wait. See the logic and commentary in `pushHeader`.

Ever since 186e9a8a, we also need to include a `ClockWhilePending` observation to advance the pushed intent's `local_timestamp`. If we don't advance the `local_timestamp` then it will still be seen as uncertain even if its timestamp is above a reader's local uncertainty limit.

So there is a solution here along the lines that you're mentioning, but it comes with trade-offs and likely some risk due to the new interactions. Given that this is primarily for BACKUP and we'd like to rework all of this soon (in https://github.com/cockroachdb/cockroach/issues/94730), I decided to keep this simple but limited instead of more complete, but complex, error-prone, and slower to backport.
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1722 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104784#-NY3iqy63FCmqk6HWoe3:-NYJpgFU61phV7uine_3:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/a89775b23851a8358b668006fd176dce1f3fd435/pkg/kv/kvserver/concurrency/lock_table.go#L1722)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Noting that once non-locking readers stop blocking on `Exclusive` locks, we should be able to guarantee that there is an intent on the key if we ever hit this fast path. At that point, we should be able to assert that the lock is not only held as unreplicated. 

Maybe worth capturing in a TODO, while we have all this paged in?
</blockquote></details>

Good point, done.
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1724 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104784#-NY3nsFa439QdcLMWwlJ:-NYK-DRN6kFjb9V0CfCq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a89775b23851a8358b668006fd176dce1f3fd435/pkg/kv/kvserver/concurrency/lock_table.go#L1724)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Following on from our conversation yesterday, where we decided we wouldn't call `tryUpdateLockLocked` from `resumeScan` for pushed transactions (we'll only do so for finalized transactions). 

I don't think we want the same behaviour for unreplicated locks -- that's because they don't trigger resolve intent requests, so they won't be calling into `UpdateLocks` as a result. Mostly a note to myself that this should compose well with https://github.com/cockroachdb/cockroach/pull/104782. 
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 2858 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104784#-NY3jZWb4GIJqw8BYYIi:-NYJq9wmB2yAdMW7WS2A:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a89775b23851a8358b668006fd176dce1f3fd435/pkg/kv/kvserver/concurrency/lock_table.go#L2858)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Consider adding a comment that explains this `g.hasUncertaintyInterval` check (or points to the comment in `tryActiveWait`).
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/concurrency/testdata/lock_table/resolve_pushed_txn_locks` line 125 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104784#-NY4-Ah886R45ODeAtnE:-NYJqOBZCXkKHuZNBSA2:blxb8h) ([raw file](https://github.com/cockroachdb/cockroach/blob/a89775b23851a8358b668006fd176dce1f3fd435/pkg/kv/kvserver/concurrency/testdata/lock_table/resolve_pushed_txn_locks#L125)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Once we implement the changes to not call into `tryUpdateLock` from `resumeScan` for lock updates that correspond to unfinalized transactions, we should see the lock on c back here as well, right?
</blockquote></details>

Right, this has now changed.


<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
bors r=arulajmani
